ompi/mca/coll/han/coll_han_scatterv.c

```diff
@@ -0,0 +1,388 @@
+            ptrdiff_t offset = 0;
+
+            for (int i = 0; i < w_size; ++i) {
+                up_peer = topo[2 * i];
+                if (root_up_rank == up_peer) {
+                    continue;
+                }
+
+                w_peer = topo[2 * i + 1];
+
+                ompi_datatype_copy_content_same_ddt(sdtype, (size_t) scounts[w_peer],
+                                                    bounce_buf + offset,
+                                                    (char *) sbuf
+                                                        + (size_t) displs[w_peer] * sdsize);
+                offset += sdsize * (size_t) scounts[w_peer];
+            }
+
+            reorder_sbuf = bounce_buf;
+        }
+
+        /* Up Iscatterv */
+        up_comm->c_coll->coll_iscatterv((const char *) reorder_sbuf, up_scounts, up_displs, sdtype,
+                                        rbuf, rcount, rdtype, root_up_rank, up_comm, &iscatterv_req,
+                                        up_comm->c_coll->coll_scatterv_module);
+
+        /* Low Scatterv */
+        low_comm->c_coll->coll_scatterv(sbuf, low_scounts, low_displs, sdtype, rbuf, rcount, rdtype,
+                                        root_low_rank, low_comm,
+                                        low_comm->c_coll->coll_scatterv_module);
+
+        ompi_request_wait(&iscatterv_req, MPI_STATUS_IGNORE);
+
+    root_out:
+        if (low_displs) {
+            free(low_displs);
+        }
+        if (low_scounts) {
+            free(low_scounts);
+        }
+        if (up_displs) {
+            free(up_displs);
+        }
+        if (up_scounts) {
+            free(up_scounts);
+        }
+        if (up_peer_lb) {
+            free(up_peer_lb);
+        }
+        if (up_peer_ub) {
+            free(up_peer_ub);
+        }
+        if (bounce_buf) {
+            free(bounce_buf);
+        }
+        if (iscatterv_req) {
+            ompi_request_free(&iscatterv_req);
+        }
+
+        return err;
+    }
+
+    /* #################### Root's local peers ########################### */
+    if (root_up_rank == up_rank) {
+        /* Low Scatterv */
+        low_comm->c_coll->coll_scatterv(NULL, NULL, NULL, NULL, rbuf, rcount, rdtype, root_low_rank,
+                                        low_comm, low_comm->c_coll->coll_scatterv_module);
+        return OMPI_SUCCESS;
+    }
+
+    /* #################### Other node followers ########################### */
+    size_t *low_data_size = NULL, rdsize = 0, receive_size = 0;
+
+    ompi_datatype_type_size(rdtype, &rdsize);
+    receive_size = rdsize * (size_t) rcount;
+
+    if (root_low_rank != low_rank) {
+        /* Low Gather - Gather each local peer's receive data size */
+        low_comm->c_coll->coll_gather((const void *) &receive_size, 1, MPI_UINT64_T, NULL, 1,
+                                      MPI_UINT64_T, root_low_rank, low_comm,
+                                      low_comm->c_coll->coll_gather_module);
```

Comment:
Technically, this is not safe: `size_t` is a `uint64_t` only on 64bit systems. I think we dropped support for 32bit but there were people who wanted to revive that. I suggest to be consistent and use `uint64_t` for receive_size if you want to send it as `MPI_UINT64_T`

Reply:
Got it. Fixed. Thank you!